numba/types/npytypes.py

```diff
@@ -42,47 +42,113 @@ def key(self):
         return self.count
 
 
+_RecordField = collections.namedtuple('_RecordField', 'type,offset')
+
+
 class Record(Type):
     """
-    A Numpy structured scalar.  *descr* is the string representation
-    of the Numpy dtype; *fields* of mapping of field names to
-    (type, offset) tuples; *size* the bytesize of a record;
-    *aligned* whether the fields are aligned; *dtype* the Numpy dtype
-    instance.
+    A Record datatype can be mapped to a Numpy structured dtype.
+    A record is very flexible since it is layout as a list of bytes.
+    Fields can be mapped to arbitrary points inside it, even if they overlap.
+
+    *fields* is a list of `(name:str, data:dict)`.
+        data is `{ type: Type, offset: int }`
+    *size* is an int; the record size
+    *aligned* is a boolean; whether the record is ABI aligned.
     """
     mutable = True
 
-    def __init__(self, descr, fields, size, aligned, dtype):
-        self.descr = descr
-        self.fields = fields.copy()
+    @classmethod
+    def make_c_struct(cls, name_types):
+        """Construct a Record type from a list of (name:str, type:Types).
+        The layout of the structure will follow C.
+
+        Note: only scalar types are supported currently.
+        """
+        from numba.targets.registry import cpu_target
+
+        ctx = cpu_target.target_context
+        offset = 0
+        fields = []
+        for k, ty in name_types:
+            if not isinstance(ty, Number):
+                raise TypeError("invalid type specified: {}".format(ty))
```

Comment:
Perhaps this error message could also contain the reason why it is invalid `(only scalar (types.Number) types are supported)`?

Reply:
done.  I will need to change it in the later PR for added support of NestedArray.